src/lib/pubkey/ec_group/ec_inner_pc.cpp

```diff
@@ -0,0 +1,209 @@
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v - checked_ref(other).value());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::mul(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v * checked_ref(other).value());
+}
+
+void EC_Scalar_Data_PC::serialize_to(std::span<uint8_t> bytes) const {
+   BOTAN_ARG_CHECK(bytes.size() == m_group->order_bytes(), "Invalid output length");
+   m_group->pcurve().serialize_scalar(bytes, m_v);
+}
+
+EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
+                                               PCurve::PrimeOrderCurve::AffinePoint pt) :
+      m_group(std::move(group)),
+      m_pt(std::move(pt)),
+      m_bytes(m_pt.serialize<secure_vector<uint8_t>>()),
+      m_x_bytes(std::span{m_bytes}.subspan(1, field_element_bytes())),
+      m_y_bytes(std::span{m_bytes}.last(field_element_bytes())) {}
+
+EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
+                                               std::span<const uint8_t> bytes) :
+      m_group(std::move(group)),
+      m_pt(deserialize_pcurve_pt(m_group->pcurve(), bytes)),
+      m_bytes(m_pt.serialize<secure_vector<uint8_t>>()),
+      m_x_bytes(std::span{m_bytes}.subspan(1, field_element_bytes())),
+      m_y_bytes(std::span{m_bytes}.last(field_element_bytes())) {}
+
+const EC_AffinePoint_Data_PC& EC_AffinePoint_Data_PC::checked_ref(const EC_AffinePoint_Data& data) {
+   const auto* p = dynamic_cast<const EC_AffinePoint_Data_PC*>(&data);
+   if(!p) {
+      throw Invalid_State("Failed conversion to EC_AffinePoint_Data_PC");
+   }
+   return *p;
+}
+
+std::unique_ptr<EC_AffinePoint_Data> EC_AffinePoint_Data_PC::clone() const {
+   return std::make_unique<EC_AffinePoint_Data_PC>(m_group, m_pt);
+}
+
+const std::shared_ptr<const EC_Group_Data>& EC_AffinePoint_Data_PC::group() const {
+   return m_group;
+}
+
+std::unique_ptr<EC_AffinePoint_Data> EC_AffinePoint_Data_PC::mul(const EC_Scalar_Data& scalar,
+                                                                 RandomNumberGenerator& rng,
+                                                                 std::vector<BigInt>& ws) const {
+   BOTAN_UNUSED(ws);
+
+   BOTAN_ARG_CHECK(scalar.group() == m_group, "Curve mismatch");
+   const auto& k = EC_Scalar_Data_PC::checked_ref(scalar).value();
+   auto pt = m_group->pcurve().mul(m_pt, k, rng).to_affine();
+   return std::make_unique<EC_AffinePoint_Data_PC>(m_group, std::move(pt));
+}
+
+size_t EC_AffinePoint_Data_PC::field_element_bytes() const {
+   return m_group->pcurve().field_element_bytes();
+}
+
+void EC_AffinePoint_Data_PC::serialize_x_to(std::span<uint8_t> bytes) const {
+   const size_t fe_bytes = this->field_element_bytes();
+   BOTAN_ARG_CHECK(bytes.size() == fe_bytes, "Invalid output size");
+   copy_mem(bytes, m_x_bytes);
```

Comment:
Nit: `copy_mem` does that size check internally, same in some methods that follow. Obviously, with a less clear error message.
Please bear with me if I did point that out before and you decided against it. 😓

Reply:
Intentional for the error message - this function is (at least through several layers) reachable by user code, and while they could probably figure it out just from a generic error message it seemed nicer to be a bit more explicit. TBH I think of `copy_mem`s check as more of a fallback for cases we forget.